pkg/littlefs2/fs/littlefs2_fs.c

```diff
@@ -62,29 +62,35 @@ static int littlefs_err_to_errno(ssize_t err)
 }
 
 static int _dev_read(const struct lfs_config *c, lfs_block_t block,
-                 lfs_off_t off, void *buffer, lfs_size_t size)
+                     lfs_off_t off, void *buffer, lfs_size_t size)
 {
     littlefs2_desc_t *fs = c->context;
     mtd_dev_t *mtd = fs->dev;
 
     DEBUG("lfs_read: c=%p, block=%" PRIu32 ", off=%" PRIu32 ", buf=%p, size=%" PRIu32 "\n",
           (void *)c, block, off, buffer, size);
 
-    return mtd_read_page(mtd, buffer, (fs->base_addr + block) * mtd->pages_per_sector,
-                         off, size);
+    uint32_t page = (fs->base_addr + block) * fs->sectors_per_block * mtd->pages_per_sector;
+    page += off / mtd->page_size;
+    off &= (mtd->page_size - 1); /* modulo power of two */
```

Comment:
Otherwise if the config is there to save some instructions on every read / write in the common case, you want to also wrap this inside a `if (CONFIG_LITTLEFS2_MIN_BLOCK_SIZE_EXP > 0)`

Reply:
I think the block to page translation must be there anyway, because even if `sectors_per_block = 1` then block size may be unequal to page size, which means reading a block at some offset within the block may be some page at another offset. But there must be something, not visible in the function, that makes it work somehow. Else I think someone would have noticed wrong data being delivered. And even if you say that we only read/write full blocks, this would raise the question why there is an `off` parameter.

Reply:
That's why the fs block is multiplied with `pages_per_sector` to get the mtd page (assuming 1 block = 1 sector)

Reply:
For example: https://github.com/RIOT-OS/RIOT/blob/1638606c5c01b24d967e5630fa43efad5437add9/boards/same54-xpro/board.c#L46-L53
The sector size is `256 * 16 = 4096`

Consider `_dev_write()` with `block = 110` and `off = 300`.
https://github.com/RIOT-OS/RIOT/blob/1638606c5c01b24d967e5630fa43efad5437add9/pkg/littlefs2/fs/littlefs2_fs.c#L77-L88

The calculated page is `110 * 16 = 1760`, but  the offset is greater then the page size, which is not handled in the function. Or what am I missing?

Reply:
> which is not handled in the function

But [it is](https://github.com/RIOT-OS/RIOT/blob/master/drivers/mtd/mtd.c#L267) - contrary to what the documentation claims unfortunately :weary:

Reply:
Ah ok. then I can remove these lines and update the doc